### PR TITLE
[Merged by Bors] - Extract Resources into their own dedicated storage

### PR DESCRIPTION
**Pull request by james7132**
# Objective
At least partially addresses #6282.

Resources are currently stored as a dedicated Resource archetype (ID 1). This allows for easy code reusability, but unnecessarily adds 72 bytes (on 64-bit systems) to the struct that is only used for that one archetype. It also requires several fields to be `pub(crate)` which isn't ideal.

This should also remove one sparse-set lookup from fetching, inserting, and removing resources from a `World`.

## Solution

- Add `Resources` parallel to `Tables` and `SparseSets` and extract the functionality used by `Archetype` in it.
- Remove `unique_components` from `Archetype`
- Remove the `pub(crate)` on `Archetype::components`.
- Remove `ArchetypeId::RESOURCE`
- Remove `Archetypes::resource` and `Archetypes::resource_mut`

---

## Changelog
Added: `Resources` type to store resources.
Added: `Storages::resource`
Removed: `ArchetypeId::RESOURCE`
Removed: `Archetypes::resource` and `Archetypes::resources`
Removed: `Archetype::unique_components` and `Archetypes::unique_components_mut`

## Migration Guide
Resources have been moved to `Resources` under `Storages` in `World`. All code dependent on `Archetype::unique_components(_mut)` should access it via `world.storages().resources()` instead.

All APIs accessing the raw data of individual resources (mutable *and* read-only) have been removed as these APIs allowed for unsound unsafe code. All usages of these APIs should be changed to use `World::{get, insert, remove}_resource`.

### CODE REVIEW

File: crates/bevy_ecs/src/world/world_cell.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/world_cell.rs:186]: This bit of logic is duplicated repeatedly. Should we make an internal function for it?

File: crates/bevy_ecs/src/storage/resource.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/resource.rs:7]: This needs a doc string; it's pub and relatively easy to stumble across.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Very happy with the direction here; this is much cleaner.

A couple of small changes needed.

1. Finish the migration guide.
2. Add doc strings to `Resources`.


Optionally: 

1. Refactor the common snippet noted above into a function.
2. Add doc strings to the methods on `Resources`.

File: crates/bevy_ecs/src/world/world_cell.rs
Comment (james7132) [reply crates/bevy_ecs/src/world/world_cell.rs:186]: Done.

**Review by alice-i-cecile — APPROVED** [review]
@bevyengine/ecs-team This one is easy; reviews?

**Comment by alice-i-cecile** [comment]
@james7132 can you fix this up when you get the chance? I think the fixes should be straightforward, and I'd like to unblock #4955.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Ping @BoxyUwU for attempt 5 to remember to review this 😂

File: crates/bevy_ecs/src/world/mod.rs
Comment (BoxyUwU) [inline crates/bevy_ecs/src/world/mod.rs:1198]:
```suggestion
        let resources = &mut self.storages.resources.resources;
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (BoxyUwU) [inline crates/bevy_ecs/src/world/mod.rs:1334]:
```suggestion
        // SAFE: This function has exclusive access to the world so nothing aliases `ticks`.
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (BoxyUwU) [inline crates/bevy_ecs/src/world/mod.rs:1342]:
```suggestion
            // SAFE: This function has exclusive access to the world so nothing aliases `ptr`.
```

**Review by BoxyUwU — CHANGES_REQUESTED** [review]
5th times the charm I suppose

File: crates/bevy_ecs/src/world/mod.rs
Comment (james7132) [reply crates/bevy_ecs/src/world/mod.rs:1198]: Oh man just realized how hard that is to read with the double resources in there.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Done.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Done.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Done.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:109]: On second thought, this should be at most `pub(crate)`, there's no way to safely use this outside of the crate. I'll fix up the safety comment to reflect the actual conditions as well.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Done.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:113]: Scratch that last comment, meant to send it for `insert`. Though I guess if `insert` is `pub(crate)`, so should `remove`?

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Done.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: After realizing that this is pretty badly structured for external consumers of this API, I reworked the overall structure of the API to make this a bit easier to work with. Users will now need to either `get_mut` or `initialize_with` a `ResourceInfo` and then write to it, with `ResourceInfo` ensuring the invariant of "zero or one" elements is maintained. This has the added benefit of removing pretty much all references of `Column` related to resources from non-storage code.

**Comment by james7132** [comment]
@alice-i-cecile @BoxyUwU after realizing inserting and removing resources from the storage API was basically impossible to use safely form outside of bevy_ecs, I reworked a lot of this to operate with `ResourceInfo` instead.

**Comment by alice-i-cecile** [comment]
@plof27 I want your review on this too.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/resource.rs:8]:
```suggestion
/// Metadata for [`Resource`] data stored in the [`World`].
pub struct ResourceInfo {
```

File: crates/bevy_ecs/src/storage/table.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/storage/table.rs:128]:
```suggestion
    #[must_use = "The returned pointer should be used to drop the removed component"]
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1136]:
```suggestion
                        panic!("No resource of type {} exists in the World.", std::any::type_name::<R>())
```

**Review by alice-i-cecile — APPROVED** [review]
Some small improvements, but nothing blocking. This is a good architectural change, and like usual, the workmanship is solid. The unsafe invariants appear to be upheld where modified.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (BoxyUwU) [reply crates/bevy_ecs/src/storage/resource.rs:8]: its not just metadata it's _also_ the actual data for the resource, that `column` is basically `Option<Box<???Component??>>`

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/storage/resource.rs:8]: Ah I see. Definitely needs to be renamed then.

Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:8]: `ResourceData`? Open to some bikeshedding here. I suck at naming types.

Comment (BoxyUwU) [reply crates/bevy_ecs/src/storage/resource.rs:8]: `ResourceData` works for me

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/storage/resource.rs:8]: Yep, I like it.

**Comment by alice-i-cecile** [comment]
Needs a Migration Guide.

**Comment by alice-i-cecile** [comment]
> [1:32 AM]Alice 🎃: Do you think this deserves the controversial label? I guess I don't really think that this is a particularly challenging PR, or something that anyone opposes
> [1:34 AM]james7132: Personally, no. Though the unsafe in it was what warranted the label initially I think.

Removing the controversial label; I don't think this is warranted. @JoJoJet @DJMcNab @TheRawMeatball I'd like an approval from one of you + Boxy as well though.

**Review by alice-i-cecile — APPROVED** [review]
Rebase LGTM, and I like the name change to `ResourceData`.

File: crates/bevy_ecs/src/archetype.rs
Comment (JoJoJet) [inline crates/bevy_ecs/src/archetype.rs:142]: I like this encapsulation.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (JoJoJet) [inline crates/bevy_ecs/src/storage/resource.rs:132]:
```suggestion
    /// Gets read-only access to a resource, if it exists.
```

Seems like this was refactored to not return a `Ptr` anymore, so the docs should be updated to reflect that.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (JoJoJet) [inline crates/bevy_ecs/src/storage/resource.rs:138]:
```suggestion
    /// Gets mutable access to a resource, if it exists.
```

File: crates/bevy_ecs/src/storage/resource.rs
Comment (JoJoJet) [inline crates/bevy_ecs/src/storage/resource.rs:12]: What's the reasoning for using a unit-length `Column`, instead of something with `Option`?

**Review by JoJoJet — APPROVED** [review]
LGTM, just a few nits + a question

File: crates/bevy_ecs/src/storage/resource.rs
Comment (BoxyUwU) [reply crates/bevy_ecs/src/storage/resource.rs:12]: the closest type for what we want is `Option<Box<dyn Component>>` + changeticks I think. unfortunately we cant actually use that because bevy supports/will support components/resources that dont have rust types which is incompatible with `dyn Component`. I assume `Column` was chosen because its the only thing we've already got implemented that manages a heap allocation without specifying a type

Comment (JoJoJet) [reply crates/bevy_ecs/src/storage/resource.rs:12]: Makes sense. Perhaps a more specialized type could be used in a future PR, if it ends up being faster.

**Review by BoxyUwU — CHANGES_REQUESTED** [review]

**Comment by BoxyUwU** [comment]
I think all the unsoundness to do with `!Send` stuff would just disappear if we made `World: !Send` hold like it ought to. I don't know whether its worth blocking this PR on the `!Send` unsoundness since you could reasonably just blame the unsound `World: Send` impl (#3310).

**Comment by BoxyUwU** [comment]
Unlike `!Send` though, resources can _also_ be `!Sync` which makes the `resource.rs` methods that let you go `&self -> Ptr<'_>` wrong for the same reason as the `!Send` unsoundness. You could probably also blame the `World: Sync` impl here but iunno, the `Send`/`Sync` on world are kind of a mess and I don't know what we're gonna do here "long term". 

Safest is probably to not merge this PR until all the methods are made unsafe with appropriate safety invariants to do with send/syncness :upside_down_face:  I am not sure if we actually have any unsoundness Today to do with `World: Sync` so we definitely cant merge this if we're adding some, unlike `World: Send` where there is at least some form of consensus that `World: Send` is wrong

**Comment by james7132** [comment]
Brainfarted here and forgot about the prior effort to make mutating accesses in bevy_ecs non-pub. `insert`/`remove` and friends should probably not be pub, and `get_mut` should be more scoped than giving direct access to ResourceData. I'll add these changes soon.

> Unlike !Send though, resources can also be !Sync which makes the resource.rs methods that let you go &self -> Ptr<'_> wrong for the same reason as the !Send unsoundness. You could probably also blame the World: Sync impl here but iunno, the Send/Sync on world are kind of a mess and I don't know what we're gonna do here "long term".
  
This is interesting since the existing `unique_components` API also currently allows this IIRC. Not that we shouldn't fix this, but pointing out that this isn't necessarily a regression.

**Comment by james7132** [comment]
Thinking on this a bit more, `Resources` within a `World` are always read-only as there is no public `World::storages_mut`. Perhaps it might be best to make all data related accesses, mut or not, `pub(crate)`.

**Comment by BoxyUwU** [comment]
I think it would be fine to make it all `pub(crate)` since we ought to be covering all the require functionality as `World` methods anyway :thinking: doesnt really change that these methods are all kinda busted with the current send/sync impls though, even if its not publicly a problem its still a huge hazard for maintaining bevy

**Comment by james7132** [comment]
There's still (some?) value in being able to get the change ticks and the ArchetypeComponentId. Is that worth exposing at all?

**Comment by alice-i-cecile** [comment]
I think we can do those in follow-up PRs. It's useful, but not essential.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:84]: Made all of the data related methods private.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:63]: Made all of the data related methods private.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:44]: Made all of the data related methods private.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Updated the safety comment.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:75]: Updated the safety comment.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs:107]: Updated the safety comment and marked it (and remove) as unsafe fns.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Looks like by making it `pub(crate)` highlighted the API as dead code. Removed.

File: crates/bevy_ecs/src/storage/resource.rs
Comment (james7132) [reply crates/bevy_ecs/src/storage/resource.rs]: Does this look better?

**Review by BoxyUwU — APPROVED** [review]
Honestly this is still kinda sketchy around `Send`/`Sync` but I am feeling like we ought to just merge this and move on because trying to get bevy to be sound with `World: Send + Sync` seems (to me) like a losing battle. Since this PR doesnt actually make anything _worse_ we should just merge this and make `World: !Send + !Sync` hold later.

**Comment by alice-i-cecile** [comment]
This is ready: I agree with Boxy's assessment and think this makes reasoning about the internals of resources much nicer. Small perf improvements are a nice bonus.

bors r+

**Pull request closed without merging**
